File: netlify/functions/save_to_db.mjs
import { getDatabase } from "@netlify/database";

export default async (req) => {
  try {
    const body = await req.json().catch(() => ({}));
    const rows = Array.isArray(body.rows) ? body.rows : [];
    const uploadDate = body.date || new Date().toISOString().slice(0, 10);

    const db = getDatabase();
    const client = await db.pool.connect();

    try {
      await client.query("BEGIN");

      for (const row of rows) {
        const ticker = String(row.ticker ?? "").trim().toUpperCase();
        if (!ticker) continue;

        const newVals = [
          Number(row.pct_change_current_quarter),
          Number(row.pct_change_next_quarter),
          Number(row.pct_change_current_fiscal_year),
          Math.trunc(Number(row.num_current_fiscal_year_estimates)),
        ];

        const existing = await client.query(
          `SELECT pct_change_current_quarter,
                  pct_change_next_quarter,
                  pct_change_current_fiscal_year,
                  num_current_fiscal_year_estimates,
                  count
             FROM tickers
            WHERE ticker = $1`,
          [ticker]
        );

        if (existing.rowCount === 0) {
          await client.query(
            `INSERT INTO tickers (
               ticker, date,
               pct_change_current_quarter,
               pct_change_next_quarter,
               pct_change_current_fiscal_year,
               num_current_fiscal_year_estimates,
               count
             ) VALUES ($1, $2, $3, $4, $5, $6, 1)`,
            [ticker, uploadDate, ...newVals]
          );
        } else {
          const e = existing.rows[0];
          const oldVals = [
            Number(e.pct_change_current_quarter),
            Number(e.pct_change_next_quarter),
            Number(e.pct_change_current_fiscal_year),
            Number(e.num_current_fiscal_year_estimates),
          ];
          const oldCount = Number(e.count);

          let increases = 0;
          let decreases = 0;
          for (let i = 0; i < 4; i++) {
            if (newVals[i] > oldVals[i]) increases++;
            else if (newVals[i] < oldVals[i]) decreases++;
          }
          const newCount = decreases === 0 && increases >= 2 ? oldCount + 1 : oldCount;

          await client.query(
            `UPDATE tickers
                SET date = $1,
                    pct_change_current_quarter = $2,
                    pct_change_next_quarter = $3,
                    pct_change_current_fiscal_year = $4,
                    num_current_fiscal_year_estimates = $5,
                    count = $6
              WHERE ticker = $7`,
            [uploadDate, ...newVals, newCount, ticker]
          );
        }
      }

      await client.query(
        `INSERT INTO uploads (upload_date, status)
         VALUES ($1, 'success')
         ON CONFLICT (upload_date)
         DO UPDATE SET status = EXCLUDED.status`,
        [uploadDate]
      );

      await client.query("COMMIT");
    } catch (err) {
      await client.query("ROLLBACK");
      throw err;
    } finally {
      client.release();
    }

    return Response.json({ message: "Saved successfully" });
  } catch (e) {
    return Response.json({ error: e.message }, { status: 500 });
  }
};

export const config = {
  path: "/.netlify/functions/save_to_db",
};

File: netlify/functions/save_to_db.py
import json
import os
import psycopg2
from datetime import date

def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
        rows = body.get("rows", [])
        upload_date = body.get("date") or str(date.today())

        conn = psycopg2.connect(os.environ["DATABASE_URL"])
        cur = conn.cursor()

        for row in rows:
            ticker = row["ticker"].strip().upper()
            new_vals = [
                float(row["pct_change_current_quarter"]),
                float(row["pct_change_next_quarter"]),
                float(row["pct_change_current_fiscal_year"]),
                int(row["num_current_fiscal_year_estimates"])
            ]

            cur.execute("""
                SELECT pct_change_current_quarter,
                       pct_change_next_quarter,
                       pct_change_current_fiscal_year,
                       num_current_fiscal_year_estimates,
                       count
                FROM tickers
                WHERE ticker = %s
            """, (ticker,))
            existing = cur.fetchone()

            if existing is None:
                cur.execute("""
                    INSERT INTO tickers (
                        ticker, date,
                        pct_change_current_quarter,
                        pct_change_next_quarter,
                        pct_change_current_fiscal_year,
                        num_current_fiscal_year_estimates,
                        count
                    ) VALUES (%s, %s, %s, %s, %s, %s, 1)
                """, (ticker, upload_date, *new_vals))
            else:
                old_vals = [float(existing[0]), float(existing[1]), float(existing[2]), int(existing[3])]
                old_count = int(existing[4])
                increases = sum(1 for o, n in zip(old_vals, new_vals) if n > o)
                decreases = sum(1 for o, n in zip(old_vals, new_vals) if n < o)
                new_count = old_count + 1 if decreases == 0 and increases >= 2 else old_count

                cur.execute("""
                    UPDATE tickers
                    SET date = %s,
                        pct_change_current_quarter = %s,
                        pct_change_next_quarter = %s,
                        pct_change_current_fiscal_year = %s,
                        num_current_fiscal_year_estimates = %s,
                        count = %s
                    WHERE ticker = %s
                """, (upload_date, *new_vals, new_count, ticker))

        cur.execute("""
            INSERT INTO uploads (upload_date, status)
            VALUES (%s, 'success')
            ON CONFLICT (upload_date)
            DO UPDATE SET status = EXCLUDED.status
        """, (upload_date,))

        conn.commit()
        cur.close()
        conn.close()
        return {"statusCode": 200, "body": json.dumps({"message": "Saved successfully"})}
    except Exception as e:
        return {"statusCode": 500, "body": json.dumps({"error": str(e)})}
